org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/internal/GradleScopeUtils.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2017 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.buildship.core.launch.internal;
+
+import java.util.Collections;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSet.Builder;
+import com.google.common.collect.Sets;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.jdt.core.IClasspathAttribute;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+
+import org.eclipse.buildship.core.CorePlugin;
+
+/**
+ * Contains utility methods to collect Gradle scopes.
+ *
+ * @author Donat Csikos
+ */
+public final class GradleScopeUtils {
+
+    private GradleScopeUtils() {
+    }
+
+    /**
+     * Collects the dependency scopes for the target launch configuration.
+     * <p/>
+     * An empty set is returned if
+     * <ul>
+     * <li>the scopes cannot be obtained (e.g. the target configuration is not supported),</li>
+     * <li>the current Gradle version doesn't provide scope information, or</li>
+     * <li>an exception is thrown</li>
+     * </ul>
+     * .
+     *
+     * @param configuration the target launch configuration
+     * @return the source set names
+     */
+    public static Set<String> collectScopes(ILaunchConfiguration configuration) {
```

Comment:
❌ What if the user manually added a source folder to the model, but forgot scope information and we also have other source folders with scope information? Currently this code assumes such a source folder needs no additional dependencies. I think instead we should treat this situation as "we don't know, let's put everything on the classpath".

Reply:
This makes sense. I've adjusted the code accordingly.